### ISSUE TITLE
Outlook Add-in SSO - .Net core Redirect URI not able to set and fails in Outllook Browser App

### ISSUE THREAD

**Issue by Tsathyapriya80** [OfficeDev/office-js#2960]
## Post how-to questions about your code on Microsoft Q&A

Outlook add-in with SSO using Microsoft Identity Platform. localhost -web page is working from browser, and it works on Outlook app in windows. But the same is authenticating from Outlook web app - it keeps on redirecting - looping and fails. it will be great if we can discuss some debugging tips. I have tried with Fiddler and from Developer tools, but I am missing something, or the error is not intuitive.
Outlook 365 app in windows - plugin
o When plugin opened up from Outlook App – redirected https://locahost:****/page with 302
o Redirected to login.microsftonline.com to authorize client and received 200 status, then redirected to call back path
o https://locahost:****/signin-oidc – signed in and redirected to Message read page
o https://locahost:****/page loads with 200 status
Outlook 365 Web App from browser - plugin
o When plugin opened up from Outlook App – redirected https://locahost:****/page with 302
o Redirected to login.microsftonline.com to authorize client and received 200 status, then redirected to call back path
o https://locahost:****/signin-oidc – signed in and redirected to Message read page
o https://locahost:****/page redirects -302 and go to login.microsftonline.com - authorize client and loops again and again and then fails

AzureAd": {
    "Instance": "https://login.microsoftonline.com/",
    "Domain": "[Enter the domain of your tenant, e.g. contoso.onmicrosoft.com]",
    "ClientId": "******-a1dd0d",
    "TenantId": "**********3e239671",
    "ClientSecret": "*********************",
    "ClientCertificates": [
    ],
    "CallbackPath": "/signin-oidc"
   //"CallbackPath": "/MessageRead"
  },
  "DownstreamApi": {
    /*
     'Scopes' contains space separated scopes of the Web API you want to call. This can be:
      - a scope for a V2 application (for instance api:b3682cc7-8b30-4bd2-aaba-080c6bf0fd31/access_as_user)
      - a scope corresponding to a V1 application (for instance <App ID URI>/.default, where  <App ID URI> is the
        App ID URI of a legacy v1 Web application
      Applications are registered in the https:portal.azure.com portal.
    */
    "BaseUrl": "https://graph.microsoft.com/v1.0",
    "Scopes": [
      "User.Read",
      "openid",
      "profile",
      "email",
      "offline_access"
      //  ,
      //"access_as_user"
    ]
  },
services.AddAuthentication(OpenIdConnectDefaults.AuthenticationScheme)
                        .AddMicrosoftIdentityWebApp(Configuration.GetSection("AzureAd"))
                      .EnableTokenAcquisitionToCallDownstreamApi(initialScopes)
                      .AddMicrosoftGraph(Configuration.GetSection("DownstreamApi"))
                      .AddInMemoryTokenCaches();

Reproduction Steps
Outlook 365 Web App from browser - plugin
o When plugin opened up from Outlook App – redirected https://locahost:****/page with 302
o Redirected to login.microsftonline.com to authorize client and received 200 status, then redirected to call back path
o https://locahost:****/signin-oidc – signed in and redirected to Message read page
o https://locahost:****/page redirects -302 and go to login.microsftonline.com - authorize client and loops again and again and then fails

Expected Behavior
it should open the razor page in Outlook task page
An unhandled exception occurred while processing the request.
MsalServiceException: A configuration issue is preventing authentication - check the error message from the server for details. You can modify the configuration in the application registration portal.
See https://aka.ms/msal-net-invalid-client for details. Original exception: AADSTS500112: The reply address 'https://localhost:52808/AzureADAuth/Authorize' does not match the reply address 'https://localhost:52808/signin-oidc' provided when requesting Authorization code.
Trace ID: db011eb8-e772-4f17-9a2b-8c42436c7000
Correlation ID: 14286004-5433-4fd2-9443-f1ca2b5c1a44
Timestamp: 2022-11-02 02:04:45Z
Microsoft.Identity.Client.OAuth2.OAuth2Client.ThrowServerException(HttpResponse response, RequestContext requestContext)

Exception: An error was encountered while handling the remote login.
Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler.HandleRequestAsync()

Stack Query Cookies Headers Routing
MsalServiceException: A configuration issue is preventing authentication - check the error message from the server for details. You can modify the configuration in the application registration portal. See https://aka.ms/msal-net-invalid-client for details. Original exception: AADSTS500112: The reply address 'https://localhost:52808/AzureADAuth/Authorize' does not match the reply address 'https://localhost:52808/signin-oidc' provided when requesting Authorization code. Trace ID: db011eb8-e772-4f17-9a2b-8c42436c7000 Correlation ID: 14286004-5433-4fd2-9443-f1ca2b5c1a44 Timestamp: 2022-11-02 02:04:45Z
Microsoft.Identity.Client.OAuth2.OAuth2Client.ThrowServerException(HttpResponse response, RequestContext requestContext)
Microsoft.Identity.Client.OAuth2.OAuth2Client.CreateResponse(HttpResponse response, RequestContext requestContext)
Microsoft.Identity.Client.OAuth2.OAuth2Client.ExecuteRequestAsync(Uri endPoint, HttpMethod method, RequestContext requestContext, bool expectErrorsOn200OK, bool addCommonHeaders)
Microsoft.Identity.Client.OAuth2.OAuth2Client.GetTokenAsync(Uri endPoint, RequestContext requestContext, bool addCommonHeaders)
Microsoft.Identity.Client.OAuth2.TokenClient.SendHttpAndClearTelemetryAsync(string tokenEndpoint)
Microsoft.Identity.Client.OAuth2.TokenClient.SendHttpAndClearTelemetryAsync(string tokenEndpoint)
Microsoft.Identity.Client.OAuth2.TokenClient.SendTokenRequestAsync(IDictionary<string, string> additionalBodyParameters, string scopeOverride, string tokenEndpointOverride, CancellationToken cancellationToken)
Microsoft.Identity.Client.Internal.Requests.RequestBase.SendTokenRequestAsync(string tokenEndpoint, IDictionary<string, string> additionalBodyParameters, CancellationToken cancellationToken)
Microsoft.Identity.Client.Internal.Requests.ConfidentialAuthCodeRequest.ExecuteAsync(CancellationToken cancellationToken)
Microsoft.Identity.Client.Internal.Requests.RequestBase.RunAsync(CancellationToken cancellationToken)
Microsoft.Identity.Client.ApiConfig.Executors.ConfidentialClientExecutor.ExecuteAsync(AcquireTokenCommonParameters commonParameters, AcquireTokenByAuthorizationCodeParameters authorizationCodeParameters, CancellationToken cancellationToken)
Microsoft.Identity.Web.TokenAcquisition.AddAccountToCacheFromAuthorizationCodeAsync(AuthorizationCodeReceivedContext context, IEnumerable scopes)
Microsoft.Identity.Web.MicrosoftIdentityWebAppAuthenticationBuilder+<>c__DisplayClass10_1+<b__1>d.MoveNext()
Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler.RunAuthorizationCodeReceivedEventAsync(OpenIdConnectMessage authorizationResponse, ClaimsPrincipal user, AuthenticationProperties properties, JwtSecurityToken jwt)
Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler.HandleRemoteAuthenticateAsync()

Show raw exception details
Exception: An error was encountered while handling the remote login.
Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler.HandleRequestAsync()
Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
Microsoft.AspNetCore.ResponseCompression.ResponseCompressionMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)

Show raw exception details
We have several forums where you can ask the developer community questions about your code. Try one of the following forums to ask technical questions about:
- [Developing Office Add-ins for Word, Excel, PowerPoint, Outlook, Project, Visio, and OneNote](https://docs.microsoft.com/answers/topics/office-addins-dev.html)
- [Access programming in Office](https://docs.microsoft.com/answers/topics/office-access-dev.html)
- [Developing COM or VSTO Add-ins for Office](https://docs.microsoft.com/answers/topics/office-vsto-com-dev.html)
- [Office JavaScript API](https://docs.microsoft.com/answers/topics/office-js-dev.html)
- [SharePoint and OneDrive development](https://docs.microsoft.com/answers/topics/sharepoint-dev.html)
- [Deploying and publishing Office or SharePoint apps to AppSource](https://docs.microsoft.com/answers/topics/microsoft-365-apps-publishing-dev.html)

For additional information about JavaScript programming, see the [JavaScript at Microsoft](https://docs.microsoft.com/javascript/) community hub.

**Comment by davidchesnut** [comment]
Hi @Tsathyapriya80,

Just to confirm, is this problem occurring from Office SSO (calling getAccessToken)? Or is this the fallback code for when SSO fails? This looks like a fallback implementation.

This error is informative: "The reply address 'https://localhost:52808/AzureADAuth/Authorize' does not match the reply address 'https://localhost:52808/signin-oidc' provided when requesting Authorization code." I'd search the code and app registration redirect uri to be sure the 'https://localhost:52808/AzureADAuth/Authorize' is not being used anywhere. The code always needs to match the registered redirect URI.

Using Fiddler and dev tools is a good debugging approach. Finding the identity errors like AADSTS500112 which you already did is the most helpful. Another tool you may find useful is https://jwt.ms/ in which you paste the raw text of a token and examine claims and other information when you suspect a token is not configured as expected.

Cheers,
David

**Comment by davidchesnut** [comment]
BTW: For more debugging help, you can also find more info on logging and error handling in the MSAL docs: https://learn.microsoft.com/en-us/azure/active-directory/develop/msal-logging-dotnet

**Comment by Tsathyapriya80** [comment]
Hi David, I am not getting access token using JS. it does in the middleware. So I added the Redirect URI as it requires for Outlook web app

**Comment by Tsathyapriya80** [comment]
In Startup --- 
            var provider = services.BuildServiceProvider();
            var initialScopes = Configuration.GetValue<string>("DownstreamApi:Scopes")?.Split(' ');

            services.AddAuthentication(OpenIdConnectDefaults.AuthenticationScheme)
                        .AddMicrosoftIdentityWebApp(Configuration.GetSection("AzureAd"))
                      .EnableTokenAcquisitionToCallDownstreamApi(initialScopes)
                      .AddMicrosoftGraph(Configuration.GetSection("DownstreamApi"))
                      .AddInMemoryTokenCaches();

In Controller:

[Authorize]

    public class HomeController : BaseClientController<DefaultAppSettings>
    {
             private readonly ILogger<HomeController> _logger;

        private readonly GraphServiceClient _graphServiceClient;
      
        public HomeController(ILogger<HomeController> logger,
                         GraphServiceClient graphServiceClient)
        {
            _logger = logger;
            _graphServiceClient = graphServiceClient;
           
        }

        [AuthorizeForScopes(ScopeKeySection = "DownstreamApi:Scopes")]
        public async Task<IActionResult> Index()
        {
            var user = await _graphServiceClient.Me.Request().GetAsync();
            ViewData["ApiResult"] = user.DisplayName;         
           
         
            return View();
        }

**Comment by davidchesnut** [comment]
@mattgeim can you help with this issue? Thanks!

**Comment by davidchesnut** [comment]
Hi @Tsathyapriya80 still working with the team to provide some guidance here. Hope to have an answer soon.

**Comment by davidchesnut** [comment]
Hi @Tsathyapriya80, Okay providing some more information from the identity team here:

It's very important you implement the correct sequence of flow for auth sign-in to work correctly. If you are implementing SSO you want to use the [getAccessToken](https://learn.microsoft.com/javascript/api/office/office.auth#office-office-auth-getaccesstoken-member(1)) API to acquire the token. The following diagram shows the sequence of events to use the token to call your REST API and then call Microsoft Graph via the On-Behalf-Of (OB) flow

![SSO sequence diagram](https://raw.githubusercontent.com/davidchesnut/github-shared-content/main/assets/sso-sequence-diagram.png)

When SSO fails, or is unavailable, you will use a fallback mechanism through MSAL and it looks like your code is currently doing this. From the client side (task pane) you need to pop up the auth sign in in a separate browser window using [displayDialogAsync](https://learn.microsoft.com/javascript/api/office/office.ui). You'll need to register an SPA callback (in addition to the web callback) in your Azure app registration. The following sequence diagram shows the interaction that occurs between your add-in, the popup dialog you create, your middle-tier ASP.NET server, and Azure/identity.

![Fallback auth sequence diagram](https://raw.githubusercontent.com/davidchesnut/github-shared-content/main/assets/fallback-auth-sequence-diagram.png)

You can find more details on SSO and fallback flows at [Enable single sign-on (SSO) in an Office Add-in](https://learn.microsoft.com/office/dev/add-ins/develop/sso-in-office-add-ins)

Cheers,
David

**Comment by Tsathyapriya80** [comment]
Thank you, David, I will add the JS to retrieve "getaccesstoken" and implement the pop up to sign in. if your team has any sample code with .NET core and SSO for Outlook add-in could you please check-in to the github.

Thanks,
Priya

**Comment by Tsathyapriya80** [comment]
Hi David,
In the above flow, accesstoken is retrieved from office.js and goes to App Middle-tier, Is this retrieving the accesstoken in the *.cshtml - by adding javascript?

**Comment by davidchesnut** [comment]
You can refer to this sample for ASP.NET. It's not Outlook, but the pattern is the same. https://github.com/OfficeDev/Office-Add-in-samples/tree/main/Samples/auth/Office-Add-in-ASPNET-SSO

There is also an Outlook sample, but it is out of date and will be updated soon. So don't use that one for now.

Yes, you need to add the javascript file to the .cshtml file. For example, I used a Scripts section in the index.cshtml file here to load the JS files I wanted containing the auth code: https://github.com/OfficeDev/Office-Add-in-samples/blob/8cef41839eede770e0143c70f4c896dfc9e39269/Samples/auth/Office-Add-in-ASPNET-SSO/Complete/Office-Add-in-ASPNETCore-WebAPI/Views/Home/Index.cshtml#LL5C19-L5C19

Cheers,
David

**Comment by Tsathyapriya80** [comment]
Hi David,

 I have added popup and getaccesstoken. so here do we need to land into HomeController without authorization(anonymous). For my case I have added Authorize attribute. In Outlook windows app I can see context.user.claims present and it is authenticated and landing to HomeController -> Index. but for Outlook web app It doesnt contain claims and it is failing before landing to index where the displaydialogasync is given.
